### PR TITLE
[SYCL][CUDA][libclc] Add bf16 builtins and optimize half builtins for fma, fmin, fmax and fmax

### PR DESCRIPTION
**Pull request by t4c1**
For functions fma, fmin, fmax and fmax adds bf16 builtins to libclc and optimizes half builtins to use half instructions if supported by the device.

This PR also contains some changes (everything in clang folder) that have been merged in upstream llvm since last pulldown and are required for building it. There are parts of (something went wrong when merging these, so only parts were merged at first. The changes in this PR are the remainder): https://reviews.llvm.org/D118977 https://reviews.llvm.org/D117887 https://reviews.llvm.org/D119157

Tests for half changes are in https://github.com/intel/llvm-test-suite/pull/880. Tests for bf16 implementations will be added together with adding support for these to runtime in future PRs.

### CODE REVIEW

File: libclc/ptx-nvidiacl/libspirv/math/fabs.cl [hunk 1]
```diff
@@ -11,7 +11,27 @@
...
+_CLC_UNARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, ushort, __clc_fabs, ushort)
+
+_CLC_DEF _CLC_OVERLOAD uint __clc_fabs(uint x) {
+  if (__clc_nvvm_reflect_arch() >= 800) {
+    return __nvvm_abs_bf16x2(x);
+  }
+  __builtin_trap();
+  __builtin_unreachable();
+}
+_CLC_UNARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, uint, __clc_fabs, uint)
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/math/fabs.cl:37]:
```suggestion
_CLC_UNARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, uint, __clc_fabs, uint)

```

File: libclc/ptx-nvidiacl/libspirv/math/fmax.cl [hunk 2]
```diff
@@ -6,12 +6,66 @@
...
+
+#ifdef cl_khr_fp16
+
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+
+_CLC_DEF _CLC_OVERLOAD half __spirv_ocl_fmax(half x, half y) {
+  if (__clc_nvvm_reflect_arch() >= 800) {
+    return __nvvm_fmax_f16(x, y);
+  }
+  return x > y ? x : y;
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/math/fmax.cl:40]: It doesn't seem to be correct implementation of fmax instruction.
From https://www.khronos.org/registry/SPIR-V/specs/unified1/OpenCL.ExtendedInstructionSet.100.html

> If one operand is a NaN, fmax returns the other argument.

Which means if `y` is `NaN`, it must return `x`. According to my understanding expression `x > y` returns false, but expected return value is `x`, not `y`. Right?

The same comment applies to `fmin` implementation.

File: libclc/ptx-nvidiacl/libspirv/math/fmax.cl [hunk 3]
```diff
@@ -6,12 +6,66 @@
...
-#include <math/binary_builtin.inc>
+_CLC_DEF _CLC_OVERLOAD uint __clc_fmax(uint x, uint y) {
+  if (__clc_nvvm_reflect_arch() >= 800) {
+    return __nvvm_fmax_bf16x2(x, y);
+  }
+  __builtin_trap();
+  __builtin_unreachable();
+}
+_CLC_BINARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, uint, __clc_fmax, uint,
+                      uint)
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/math/fmax.cl:71]:
```suggestion
                      uint)

```

File: libclc/ptx-nvidiacl/libspirv/math/fmin.cl [hunk 4]
```diff
@@ -6,12 +6,66 @@
...
-#include <math/binary_builtin.inc>
+_CLC_DEF _CLC_OVERLOAD uint __clc_fmin(uint x, uint y) {
+  if (__clc_nvvm_reflect_arch() >= 800) {
+    return __nvvm_fmin_bf16x2(x, y);
+  }
+  __builtin_trap();
+  __builtin_unreachable();
+}
+_CLC_BINARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, uint, __clc_fmin, uint,
+                      uint)
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/math/fmin.cl:71]:
```suggestion
                      uint)

```

File: libclc/ptx-nvidiacl/libspirv/math/native_exp2.cl [hunk 5]
```diff
@@ -14,4 +14,41 @@
...
+
+int __clc_nvvm_reflect_arch();
+
+_CLC_DEF _CLC_OVERLOAD half __spirv_ocl_native_exp2(half x) {
+  if (__clc_nvvm_reflect_arch() >= 750) {
+    return __nvvm_ex2_approx_f16(x);
+  } else {
+    float upcast = x;
+    return __spirv_ocl_native_exp2(upcast);
+  }
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/math/native_exp2.cl:29]:
```suggestion
  }
  float upcast = x;
  return __spirv_ocl_native_exp2(upcast);
```

File: libclc/ptx-nvidiacl/libspirv/math/native_exp2.cl [hunk 6]
```diff
@@ -14,4 +14,41 @@
...
+  if (__clc_nvvm_reflect_arch() >= 750) {
+    return __nvvm_ex2_approx_f16x2(x);
+  } else {
+    float upcast0 = x[0];
+    float upcast1 = x[1];
+    half2 res;
+    res.s0 = __spirv_ocl_native_exp2(upcast0);
+    res.s1 = __spirv_ocl_native_exp2(upcast1);
+    return res;
+  }
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/math/native_exp2.cl:42]:
```suggestion
  }
  float upcast0 = x[0];
  float upcast1 = x[1];
  half2 res;
  res.s0 = __spirv_ocl_native_exp2(upcast0);
  res.s1 = __spirv_ocl_native_exp2(upcast1);
  return res;
```

File: libclc/ptx-nvidiacl/libspirv/math/fmax.cl (hunk 2 above)
Comment (t4c1) [reply libclc/ptx-nvidiacl/libspirv/math/fmax.cl]: Right. I changed these back to using float implementation from libdevice.

**Review by bader — APPROVED** [review]
libclc changes look good to me.

**Review by bader — APPROVED** [review]

File: clang/test/CodeGen/builtins-nvptx-native-half-type.c [hunk 7]
```diff
@@ -0,0 +1,174 @@
+// REQUIRES: nvptx-registered-target
```
Comment (smanna12) [inline clang/test/CodeGen/builtins-nvptx-native-half-type.c:1]: Could you please add a comment what test is doing?

File: clang/test/CodeGen/builtins-nvptx.c [hunk 8]
```diff
@@ -16,6 +16,12 @@
 // RUN:   | FileCheck -check-prefix=CHECK -check-prefix=LP64 %s
 // RUN: %clang_cc1 -triple nvptx-unknown-unknown -target-cpu sm_53 \
 // RUN:   -DERROR_CHECK -fcuda-is-device -S -o /dev/null -x cuda -verify %s
+// RUN: %clang_cc1 -ffp-contract=off -triple nvptx-unknown-unknown -target-cpu sm_86 -target-feature +ptx72 \
```
Comment (smanna12) [inline clang/test/CodeGen/builtins-nvptx.c:19]: Same here?

File: clang/test/CodeGen/builtins-nvptx-native-half-type.c (hunk 7 above)
Comment (t4c1) [reply clang/test/CodeGen/builtins-nvptx-native-half-type.c:1]: This change (everything in clang folder) are already merged upstream. I just added them to this PR as they are required to build it. They will be part of the next pulldown.

Comment (smanna12) [reply clang/test/CodeGen/builtins-nvptx-native-half-type.c:1]: Thanks for the clarification.

**Comment by t4c1** [comment]
I just removed the changes to native_exp2, as that is being implemented in a slightly different way in https://github.com/intel/llvm/pull/5747.

**Review by smanna12 — APPROVED** [review]
FE changes LGTM As per comment: https://github.com/intel/llvm/pull/5724#discussion_r822643809

```This change (everything in clang folder) are already merged upstream. I just added them to this PR as they are required to build it. They will be part of the next pulldown.```

**Comment by t4c1** [comment]
Done - updated the PR description.

**Pull request merged**